### PR TITLE
[Core] Support opt-in custom logits processors with speculative decoding

### PR DESCRIPTION
**Pull request by HwangJohn**
## Purpose

This PR allows external/custom V1 logits processors to opt into speculative decoding instead of being rejected unconditionally.

This does not broadly allow arbitrary custom logits processors with speculative decoding. The default remains rejection: processors must explicitly opt in and provide the spec-decode apply hook before they are accepted.

The change adds a small explicit contract:

- `LogitsProcessor.supports_spec_decode()` defaults to `False`.
- Custom processors that opt in must implement `apply_with_spec_decode(logits, num_draft_tokens)`.
- Unsupported custom processors are rejected during config initialization, before model loading.
- Opted-in processors are applied during rejection sampling for speculative target logits.
- The optional `apply_to_speculative_draft_logits(...)` hook lets a processor affect draft-token proposal when needed.
- Optional draft hook callables are cached in `LogitsProcessors` at construction time so draft-token generation does not rediscover them on every step.
- Docs explain the spec-decode contract for custom logits processors.

## Duplicate-Work Check

Before opening this PR, I checked open PRs for `custom logits processors speculative decoding` and `logits processor spec decode`.

Related PRs/issues:

- #17799 tracks custom logits processor extensibility.
- #38467 and #38469 add spec-decode support for the built-in `LogitBiasLogitsProcessor`.
- #42802 adds spec-decode handling for the built-in `MinPLogitsProcessor`.

This PR intentionally does not add built-in `logit_bias` or `min_p` support. It is not a duplicate of those PRs; it adds the generic opt-in contract, validation path, and draft-hook caching for external/custom logits processors.

AI assistance disclosure: this PR includes code, tests, and documentation prepared with OpenAI Codex assistance. I reviewed the changed lines, kept the scope limited, and ran the validation below.

## Test Plan

Validated on DGX Spark Linux/aarch64 CUDA with the vLLM Python 3.12 development environment:

- `pre-commit run --files docs/features/custom_logitsprocs.md tests/v1/logits_processors/test_custom_offline.py tests/v1/logits_processors/test_spec_decode_custom.py vllm/config/vllm.py vllm/v1/sample/logits_processor/__init__.py vllm/v1/sample/logits_processor/interface.py vllm/v1/sample/logits_processor/state.py vllm/v1/sample/rejection_sampler.py vllm/v1/spec_decode/llm_base_proposer.py`
- `python -m pytest tests/v1/logits_processors/test_spec_decode_custom.py -q -p no:cacheprovider`
- `python -m pytest tests/v1/logits_processors/test_custom_offline.py -k 'test_rejects_custom_logitsprocs and spec_dec' -q -p no:cacheprovider`
- GPU smoke with `facebook/opt-125m`, `speculative_config={"model": "ngram", "num_speculative_tokens": 1}`, and an opted-in custom logits processor whose `apply_with_spec_decode(...)` asserts CUDA logits.
- After review-readiness docstring polish: `pre-commit run --files tests/v1/logits_processors/test_custom_offline.py`
- After review-readiness docstring polish: `python -m pytest tests/v1/logits_processors/test_custom_offline.py -k 'test_rejects_custom_logitsprocs and spec_dec' -q -p no:cacheprovider`
- After rebasing onto latest `origin/main` (`445ded18c`): reran the changed-file pre-commit and both targeted pytest commands above.

## Test Result

- `pre-commit run --files ...`: passed
- `test_spec_decode_custom.py`: `10 passed`
- `test_custom_offline.py -k 'test_rejects_custom_logitsprocs and spec_dec'`: `3 passed, 8 deselected`
- GPU smoke: `gpu_smoke_ok [6, 38, 437, 10]`
- Review-readiness docstring polish pre-commit: passed
- Review-readiness docstring polish pytest: `3 passed, 8 deselected`
- After rebase to latest `origin/main`: changed-file pre-commit passed, `test_spec_decode_custom.py` still `10 passed`, and targeted `test_custom_offline.py` still `3 passed, 8 deselected`.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/v1/sample/logits_processor/__init__.py [hunk 1]
```diff
@@ -199,14 +261,33 @@ def build_logitsprocs(
...
+        custom_instances = [
+            ctor(vllm_config, device, is_pin_memory)
+            for ctor in custom_logitproc_classes
+        ]
+        logitsprocs = LogitsProcessors(
+            itertools.chain(
+                (MinTokensLogitsProcessor(vllm_config, device, is_pin_memory),),
+                custom_instances,
+            )
+        )
```
[the remark this answers is not in the document]
Comment (HwangJohn) [reply vllm/v1/sample/logits_processor/__init__.py:279]: Thanks, addressed.

The optional `apply_to_speculative_draft_logits` lookup is now cached during `LogitsProcessors` initialization as `spec_decode_draft_logits_hooks`, and `SpecDecodeBaseProposer` uses that cached list instead of doing repeated `getattr` checks in the speculative decoding hot path.

The remaining `getattr` near this area checks for the required `apply_with_spec_decode` method during logits processor setup/validation only. It is not part of the per-draft-token hot path that this comment was concerned about.

I also added coverage for caching draft logits hooks from all logits processor buckets.

File: vllm/v1/spec_decode/llm_base_proposer.py [hunk 2]
```diff
@@ -400,6 +400,38 @@ def _greedy_sample(self, hidden_states: torch.Tensor) -> torch.Tensor:
...
+            return logits
+        for processor in sampling_metadata.logitsprocs.all:
+            apply_to_draft_logits = getattr(
+                processor,
+                "apply_to_speculative_draft_logits",
+                None,
+            )
+            if apply_to_draft_logits is not None:
+                logits = apply_to_draft_logits(logits, num_draft_tokens)
+        return logits
```
[the remark this answers is not in the document]
Comment (HwangJohn) [reply vllm/v1/spec_decode/llm_base_proposer.py:460]: Thanks, addressed together with the logits processor initialization change.

`LogitsProcessors` now precomputes `spec_decode_draft_logits_hooks` once during initialization, and this path uses that cached list directly. This removes the repeated per-processor `getattr` checks from the speculative decoding hot path.

**Comment by HwangJohn** [comment]
Gentle follow-up.

This PR appears to be blocked only by the pre-run-check gate requiring a 'ready' or 'verified' label. The Gemini review feedback has been addressed, DCO and docs checks are passing, and maintainer edits are enabled.

I validated the change on DGX Spark Linux/aarch64 CUDA with the targeted pre-commit checks, targeted pytest coverage, and a GPU smoke test as listed in the PR description.

@benchislett, since this touches speculative decoding, could you advise whether this is ready for review / a 'ready' or 'verified' label, or what changes would make it reviewable?

**Comment by HwangJohn** [comment]
Rebased this PR on current `origin/main` (`e24d1b24`) and resolved the merge conflict in `vllm/v1/spec_decode/llm_base_proposer.py`.

Conflict resolution notes:
- Preserved the upstream heterogeneous-vocab draft-logits constrain/map flow.
- Applied opt-in draft logits hooks before draft sampling.
- Re-applied the heterogeneous-vocab constraint after custom draft hooks so a hook cannot resurrect invalid draft-token logits.
- Added a unit test for that heterogeneous-vocab safety case.

Validation on DGX Spark (Linux/aarch64 CUDA):

```bash
VLLM_USE_PRECOMPILED=1 /home/nvidia/.local/bin/uv pip install --python /home/nvidia/workspace/pr/vllm/.venv/bin/python -e . --torch-backend=auto
VLLM_USE_FLASHINFER_SAMPLER=0 /home/nvidia/workspace/pr/vllm/.venv/bin/python -m pytest tests/v1/logits_processors -q -p no:cacheprovider
# 58 passed, 37 warnings
/home/nvidia/workspace/pr/vllm/.venv/bin/pre-commit run --files docs/features/custom_logitsprocs.md tests/v1/logits_processors/test_custom_offline.py tests/v1/logits_processors/test_spec_decode_custom.py vllm/config/vllm.py vllm/v1/sample/logits_processor/__init__.py vllm/v1/sample/logits_processor/interface.py vllm/v1/sample/logits_processor/state.py vllm/v1/sample/rejection_sampler.py vllm/v1/spec_decode/llm_base_proposer.py
# passed
```

I also ran a small DGX GPU smoke with `facebook/opt-125m`, ngram speculative decoding, and an opted-in custom logits processor asserting that `apply_with_spec_decode()` receives CUDA logits. It completed with `gpu_smoke_ok`.

**Comment by sachinruk** [comment]
I know its been a while but any chance we can get this merged. This is exactly the PR I need since speculative decoding is not enough to guarantee the structured output that I am after.

@HwangJohn, amazing work!
